tests/test_pd_controller.cpp

```diff
@@ -0,0 +1,226 @@
+#include "linear_feedback_controller/pd_controller.hpp"
+
+using namespace linear_feedback_controller;
+
+// Generate an array of a given type with its size deduced by the number of args
+//
+// This may be usefull when you don't want to specify by hand the number of
+// elements inside an array but want to force the value_type of the array (in
+// order to force conversion & co ...)
+template <typename ValueType, typename... T>
+constexpr auto MakeArray(T &&...values) -> std::array<ValueType, sizeof...(T)> {
+  return std::array<ValueType, sizeof...(T)>{std::forward<T>(values)...};
+}
+
+// RAII mutator, use to temporary modify a value given a reference and a tmp
+// value
+//
+// WARNING: Dangling references & co ...
+template <typename ValueType>
+struct TemporaryMutate {
+  TemporaryMutate() = delete;
+
+  template <typename T>
+  TemporaryMutate(ValueType &data, T &&tmp_value)
+      : data_(data), before_(data_) {
+    data_ = std::forward<T>(tmp_value);
+  }
+
+  virtual ~TemporaryMutate() { data_ = before_; }
+
+  constexpr auto GetOldValue() const -> const ValueType & { return before_; }
+
+ private:
+  ValueType &data_;
+  ValueType before_;
+};
+
+TEST(PdControllerTest, Ctor) {
+  EXPECT_NO_THROW({ const auto pd_ctrl = PDController(); });
+}
+
+struct Gains {
+  using IdxType = Eigen::Index;
+
+  Eigen::VectorXd p;
+  Eigen::VectorXd d;
+
+  static auto Random(IdxType p_size,
+                     std::optional<IdxType> d_size = std::nullopt) -> Gains {
+    return Gains{
+        .p = Eigen::VectorXd::Random(p_size),
+        .d = Eigen::VectorXd::Random(d_size.value_or(p_size)),
+    };
+  }
+};
+
+TEST(PdControllerTest, SetGains) {
+  auto pd_ctrl = PDController();
+
+  EXPECT_NO_THROW({ pd_ctrl.set_gains(Eigen::VectorXd{}, Eigen::VectorXd{}); });
+
+  for (const auto size : MakeArray<Gains::IdxType>(1, 3, 4, 50, 1000)) {
+    SCOPED_TRACE(::testing::Message() << "Same gains size: " << size);
+    auto requested_gains = Gains::Random(size);
+    EXPECT_NO_THROW(
+        { pd_ctrl.set_gains(requested_gains.p, requested_gains.d); });
+    // FIXME: How to test the validity ??
+
+    // Test for special double values acceptance
+    for (const auto special_value : {
+             std::numeric_limits<double>::infinity(),
+             std::numeric_limits<double>::quiet_NaN(),
+             std::numeric_limits<double>::signaling_NaN(),
+         }) {
+      SCOPED_TRACE(::testing::Message() << "Special value: " << special_value);
+
+      {
+        SCOPED_TRACE("P");
+        auto _ = TemporaryMutate{requested_gains.p(0), special_value};
+        EXPECT_NO_THROW(
+            { pd_ctrl.set_gains(requested_gains.p, requested_gains.d); });
+      }
+
+      {
+        SCOPED_TRACE("D");
+        auto _ = TemporaryMutate{requested_gains.d(0), special_value};
+        EXPECT_NO_THROW(
+            { pd_ctrl.set_gains(requested_gains.p, requested_gains.d); });
+      }
+    }
+  }
+
+  using SizePair = std::array<Gains::IdxType, 2>;
```

Comment:
Is there anything against using `std::pair<Gains::IdxType>`?

Reply:
std::pair/std::tuple are especially useful when the types used are not the same (even though I'd rather define a custom struct in order to have named arguments).

Here I'd rather use a simple array and avoid including <utility> or <tuple>.